grpc/federation/cel/url.go

```diff
@@ -0,0 +1,621 @@
+package cel
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+)
+
+const URLPackageName = "url"
+
+var (
+	URLType      = cel.ObjectType("grpc.federation.url.URL")
+	UserinfoType = cel.ObjectType("grpc.federation.url.Userinfo")
+)
+
+func (x *URL) GoURL() (url.URL, error) {
+	var user *url.Userinfo
+	if u := x.GetUser(); u != nil {
+		if u2, err := u.GoUserinfo(); err != nil {
+			return url.URL{}, err
+		} else {
+			user = u2
+		}
+	}
+
+	return url.URL{
+		Scheme:      x.GetScheme(),
+		Opaque:      x.GetOpaque(),
+		User:        user,
+		Host:        x.GetHost(),
+		Path:        x.GetPath(),
+		RawPath:     x.GetRawPath(),
+		OmitHost:    x.GetOmitHost(),
+		ForceQuery:  x.GetForceQuery(),
+		RawQuery:    x.GetRawQuery(),
+		Fragment:    x.GetFragment(),
+		RawFragment: x.GetRawFragment(),
+	}, nil
+}
+
+func (x *Userinfo) GoUserinfo() (*url.Userinfo, error) {
+	if x == nil {
+		return nil, nil
+	}
+	if x.GetPasswordSet() {
+		return url.UserPassword(x.GetUsername(), x.GetPassword()), nil
+	}
+	return url.User(x.GetUsername()), nil
+}
+
+var _ cel.SingletonLibrary = new(URLLibrary)
+
+type URLLibrary struct {
+	typeAdapter types.Adapter
+}
+
+func NewURLLibrary(typeAdapter types.Adapter) *URLLibrary {
+	return &URLLibrary{
+		typeAdapter: typeAdapter,
+	}
+}
+
+func (lib *URLLibrary) LibraryName() string {
+	return packageName(URLPackageName)
+}
+
+func createURLName(name string) string {
+	return createName(URLPackageName, name)
+}
+
+func createURLID(name string) string {
+	return createID(URLPackageName, name)
+}
+
+func (lib *URLLibrary) refToGoURLValue(v ref.Val) (url.URL, error) {
+	return v.Value().(*URL).GoURL()
+}
+
+func (lib *URLLibrary) toURLValue(v url.URL) ref.Val {
+	var userinfo *Userinfo
+	if v.User != nil {
+		password, hasPassword := v.User.Password()
+		userinfo = &Userinfo{
+			Username:    v.User.Username(),
+			Password:    password,
+			PasswordSet: hasPassword,
+		}
+	}
+
+	return lib.typeAdapter.NativeToValue(&URL{
+		Scheme:     v.Scheme,
+		Opaque:     v.Opaque,
+		User:       userinfo,
+		Host:       v.Host,
+		Path:       v.Path,
+		RawPath:    v.RawPath,
+		ForceQuery: v.ForceQuery,
+		RawQuery:   v.RawQuery,
+		Fragment:   v.Fragment,
```

Comment:
The same as https://github.com/mercari/grpc-federation/pull/244#discussion_r1777962091